### PR TITLE
Raviart-Thomas elements

### PR DESCRIPTION
**Pull request by karthichockalingam**
## Impact

Enables H(div) simulations using Raviart-Thomas elements.

Kind regards,
Karthik and Nuno
(cc @nmnobre)

### CODE REVIEW

File: framework/include/bcs/VectorDivPenaltyDirichletBC.h
Comment (lindsayad) [inline framework/include/bcs/VectorDivPenaltyDirichletBC.h:28]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
```

File: framework/src/actions/SetAdaptivityOptionsAction.C
Comment (lindsayad) [inline framework/src/actions/SetAdaptivityOptionsAction.C:165]: I thought I would need to tell you to add this here but you found it on your own. Much appreciated!

File: framework/doc/content/source/bcs/VectorDivPenaltyDirichletBC.md
Comment (lindsayad) [inline framework/doc/content/source/bcs/VectorDivPenaltyDirichletBC.md:1]: I think it would be helpful to link to the documentation for `VectorPenaltyDirichletBC` and highlight the difference between the two objects and when each should be used over the other. I can imagine a user being confused on this point

**Review by lindsayad — CHANGES_REQUESTED** [review]
Thanks for this great contribution! I have reviewed the framework portion. I think @cticenhour should review the EM portion

File: framework/include/bcs/VectorDivPenaltyDirichletBC.h
Comment (nmnobre) [reply framework/include/bcs/VectorDivPenaltyDirichletBC.h]: Done.

File: framework/src/actions/SetAdaptivityOptionsAction.C
Comment (nmnobre) [reply framework/src/actions/SetAdaptivityOptionsAction.C:165]: No worries. :)

File: framework/include/bcs/VectorDivPenaltyDirichletBC.h
Comment (nmnobre) [reply framework/include/bcs/VectorDivPenaltyDirichletBC.h]: I was about to suggest we should add a clang-tidy test, which would've caught this, https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-override.html.
But it seems work might've already started in #23614?

Comment (lindsayad) [reply framework/include/bcs/VectorDivPenaltyDirichletBC.h]: We'd have to ask @loganharbour the status of that. I'm honestly surprised our `-Wall` recipes didn't catch this (or maybe CI just didn't have enough time to get there)

Comment (nmnobre) [reply framework/include/bcs/VectorDivPenaltyDirichletBC.h]: Nah, there are multiple other occurrences throughout the code, the most notable in `VectorCurlPenaltyDirichletBC.h` which we based our implementation on. :P

Comment (nmnobre) [reply framework/include/bcs/VectorDivPenaltyDirichletBC.h]: I've added a commit to add the keyword to the headers in `framework/include/bcs` which missed it.

**Comment by lindsayad** [comment]
will need to address the EM failures

**Comment by nmnobre** [comment]
> will need to address the EM failures

Should be good now, switched KSP by Jacobi and uploaded a new set of gold files. :)

**Comment by cticenhour** [comment]
Thanks for the tag @lindsayad. I plan to give this a review today.

File: framework/src/utils/Conversion.C
Comment (roystgnr) [inline framework/src/utils/Conversion.C:416]: Not sure why this switch exists - couldn't the whole thing have been replaced with a `return Utility::enum_to_string(f);`?

**Review by roystgnr — APPROVED** [review]
I'm not seeing any obvious issues, but do wait for at least Casey to have a look too before we merge.

File: framework/doc/content/source/bcs/VectorDivPenaltyDirichletBC.md
Comment (lindsayad) [reply framework/doc/content/source/bcs/VectorDivPenaltyDirichletBC.md:1]: yep!

**Review by lindsayad — COMMENTED** [review]
It would be nice to try and increase the framework coverage a bit

File: framework/src/utils/Conversion.C
Comment (nmnobre) [reply framework/src/utils/Conversion.C]: Done.

We could also remove the quadrature-to-enum map and function in MOOSE in favour of libMesh's but that would require a minor change to the latter which is missing the CONICAL and MONOMIAL quadrature rules in its string to enum map for some reason. Are you up for it? Maybe in another PR. ^^

https://github.com/idaholab/moose/blob/0e39829646255a0ff49ba89a2de43d07b07f8aab/framework/src/utils/Conversion.C#L35-L49

**Comment by nmnobre** [comment]
> It would be nice to try and increase the framework coverage a bit

Module tests don't count towards the framework coverage, do they? :/

**Comment by lindsayad** [comment]
No, they do not. If it's a lot of trouble devising something (I know it sometimes can be), I may let it slide

**Comment by nmnobre** [comment]
> No, they do not. If it's a lot of trouble devising something (I know it sometimes can be), I may let it slide

I see for H(curl)/Nedelec, you guys created `VectorFEWave` just for testing purposes despite the redundancy.
It's funny because we chose to put the tests in the EM module because it allowed us to make the kernels as modular (pun intended) as possible, by reusing existing kernels wherever possible. But it'd seem I need to build a kernel for the full PDE anyway, I'll see what I can do. :)

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests
Comment (cticenhour) [inline modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests:1]: Considering that there is significant re-use of the phrase "The system shall be capable of modeling a div-grad problem in H(div) and in...", let's use the `requirement` and `detail` feature of the test specification system to condense these requirement details a bit. See https://github.com/idaholab/moose/blob/master/test/tests/adaptivity/dont-p-refine/tests for an example.

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests
Comment (cticenhour) [inline modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests:10]: Since these use the same issue, `issues = '#26105'` could be moved to a higher level of the tree (just under `[Tests]`) to tidy up a few more lines.

**Comment by cticenhour** [comment]
> we chose to put the tests in the EM module because it allowed us to make the kernels as modular (pun intended) as possible, by reusing existing kernels wherever possible.

This was a great choice and is exactly what we *should* do for user-facing objects and capability tests. `VectorFEWave` predates some major components of the EM module and entered into MOOSE before it was merged. I personally agree with the precedent of creating a specialized test object for coverage reasons like this, since that enables the R-T capability to be tested even when the EM module isn't turned on. A bit of redundancy in a small private object is a small price to pay to me. 😄

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i
Comment (cticenhour) [inline modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i:48]: The convention of `p` and `u` as input parameter names in the `DivField` and `GradField` objects make sense in the context of this test formulation, but I am concerned that this might be confusing in a more general sense to a user instead of something like `coupled_variable` for them. @lindsayad do you have thoughts here?

**Review by cticenhour — CHANGES_REQUESTED** [review]
Some comments to address, and a few adjustments I'd like made to the test spec file.

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i
Comment (lindsayad) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i:48]: I think something like `coupled_scalar_field` and `coupled_vector_field` or something like that would be good. I don't personally like the `u` parameter for the vector field as that is used exclusively to denote the *scalar* field in Cockburn's mixed formulation literature. Also `u` is often regarded as the `variable` the kernel is acting on in MOOSE documentation

**Comment by lindsayad** [comment]
> (Or we could count module tests towards framework coverage 😇)

I am not fundamentally opposed to this but I would defer a decision on it to @loganharbour. There would be some technical hurdles. We have a lot of templates in the framework for which we can only get meaningful coverage if we pass `-O0` but our module tests are too big generally to support `-O0` and run in a tolerable amount of time

One option is to actually put some objects in the framework instead of EM if they are general

**Comment by karthichockalingam** [comment]
> @karthichockalingam I believe @loganharbour sent you an invite to join the moose-users github group which will enable automatic testing of your pull requests

I accepted the invite. Thanks.

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests
Comment (nmnobre) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/tests:1]: Good catch. Done.

File: framework/src/utils/Conversion.C
Comment (roystgnr) [reply framework/src/utils/Conversion.C]: Yeah, I'd say a separate PR is a good idea there, especially if it'll have to wait on a libMesh update.  It might be that the `CONICAL` rules were left out of that translation on purpose, on the theory that they're mostly just useful as fallbacks for `QGauss`, but omitting an entry for `MONOMIAL` had to be an oversight, and IMHO we should have translations for every option even if we expect some to go unused.

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i
Comment (nmnobre) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i]: Done.

File: modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i
Comment (nmnobre) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i:5]: I don't think I can because `VectorFunctionReaction` is part of the EM module.

Comment (cticenhour) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i:5]: That is a generic enough object that I wouldn't mind it being moved into the framework. The only reason why it is in EM is due to the fact that the EM module was the only one that required it at the time. :)

Comment (nmnobre) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i]: Done. 26.5% -> 63.5% coverage on the new lines. Still not perfect...

Comment (nmnobre) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i]: Assuming you're happy(er) with that, how do you feel about:

1) Adding `L2_RAVIART_THOMAS` support?
2) Adding an object for Dirichlet boundary conditions on `p`, like you did in libMesh?
3) Keeping `coupled_electrostatics` as the test name. To be honest the current names under `vector_fe` aren't really very inspiring, so I couldn't decide what to change mine to...

1) and 2) could perhaps be left for another PR, up to you. :)

Comment (lindsayad) [reply modules/electromagnetics/test/tests/kernels/coupled_electrostatics/coupled_electrostatics.i]: I think we should wait on 1 and 2 for another PR. I'm totally fine with keeping the `coupled_electrostatics` name

File: test/tests/kernels/vector_fe/coupled_electrostatics.i
Comment (lindsayad) [inline test/tests/kernels/vector_fe/coupled_electrostatics.i:89]: Woa I didn't know about this input ability

**Review by lindsayad — APPROVED** [review]
LGTM

**Review by cticenhour — APPROVED** [review]
LGTM as well. Thanks @nmnobre!

**Comment by lindsayad** [comment]
Thanks guys!

**Pull request merged**
